enhancements/update/automatic-updates.md

```diff
@@ -0,0 +1,176 @@
+---
+title: neat-enhancement-idea
+authors:
+  - "@wking"
+reviewers:
+  - "@abhinavdahiya"
+  - "@crawford"
+  - "@smarterclayton"
+approvers:
+  - TBD
+creation-date: 2019-11-19
+last-updated: 2019-11-21
+status: implementable
+---
+
+# Automatic Updates
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [x] Design details are appropriately documented from clear requirements
+- [x] Test plan is defined
+- [x] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+Operator-managed updates are [one of the key benefits of OpenShift 4][architecture].
+Cluster update suggestions are distributed via [Cincinnati][cincinnati-openshift], so clusters only receive update recommendations that are appropriate for their cluster.
+While manually approving updates at a per-cluster level is possible, it should not be required.
+This enhancement adds a new `automaticUpdates` property to [the ClusterVersion `spec`][api-spec] so cluster administrators can opt in to or out of automatic updates.
+
+## Motivation
+
+Manually approving updates is tedious, doesn't scale well for administrators managing many clusters, and delays the application of potential security fixes.
+Administrators may wish to enable automatic updates for any of those or other reasons.
+Currently, administrators must implement custom polling logic to check for and apply any available updates.
+This enhancement would provide those administrators with a convenient property instead.
+
+### Goals
+
+* Make it easy to opt in to and out of automatic updates.
+
+### Non-Goals
+
+* Support calendar gating or other logic about when updates may be attempted.
+* Choose whether the installer defaults to enabling automatic updates or not for new clusters.
+
+## Proposal
+
+As proposed in [this API pull request][api-pull-request], to add a new property:
+
+```go
+// automaticUpdates enables automatic updates.
+// +optional
+AutomaticUpdates bool `json:"automaticUpdates,omitempty"`
```

Comment:
I don't think booleans are great API as they are not extensible..

having a boolean makes it so that we are saying that cvo only knows update always or just don't and it is never going to do anything else, any condition if required will be done somewhere else..

personally i think something like union discriminator https://github.com/kubernetes/enhancements/pull/926 is much better suited to express the intent no-updates, auto-updates-always, auto-updates-schedule etc..

> https://github.com/openshift/enhancements/pull/124/files#diff-8354eb4f233832a1e5afa3ffc9bea8e8R154-R160

There is an overhead trade-off that needs to taken into consideration before saying CVO doesn't do local stuff...
will a user with single or handful clusters want to run local policy-agent? or is it sane for them to do it?
will the users that want different policy for their clusters like testing/staging/prod like to run 3 different policy agents?

Does the fact that we have no policy agent at this point, and not sure how far that is or even the concept of supporting upstream that is self-signed... deter anybody from even using this... hence diminishing the value prop of the boolean.

Personally i would like clusters to support at some one usecase of automatic updates based on condition..

Reply:
> I don't think booleans are great API as they are not extensible..

I'm still not clear on what sort of extensibility you'd want.  I think there's a set of policy engines which can be distributed in external chains or run inside the CVO itself, that eventually spit out a set of recommended update targets.  Then you need a boolean knob to decide if the CVO automatically initiates an update when a target becomes available.  For example:

1. Upstream Cincinnati spits out graph A.
2. On-site policy engine ingests graph A and applies local stability filtering, removing Red-Hat-recommended update edges until they have also passed local update testing.  This engine spits out graph B.
3. CVO ingests graph B via the `upstream` URI, and applies `schedule` filtering, removing all update edges if and only if the current time lies outside the configured update windows.  This engine spits out graph C.
4. CVO ingests its own graph C and looks for available updates from the current version, storing those in `AvailableUpdates`.
5. CVO looks at `AvailableUpdates` and the `AutomaticUpdates` boolean, and decides whether to automatically set `DesiredUpdate`.

So there's still lots of room for extension.  The boolean option is just an in-cluster knob for the [existing CLI flag][0].

> ...any condition if required will be done somewhere else...

No, you could still apply policy filtering on the CVO side, independent of whether the CVO automatically updates based on the results of a local policy engine.  This is what I was trying to convey [here][1].

[0]: https://github.com/openshift/cluster-version-operator/blob/b9a3ffafdc32e534c68112a13b89cf1459e1bbd9/cmd/start.go#L30
[1]: https://github.com/openshift/enhancements/commit/568a1bd7315b38fdc23f4421ff49229cb0829e0a#diff-8354eb4f233832a1e5afa3ffc9bea8e8R159-R160